src/python_testing/TC_DEM_2_2.py

```diff
@@ -0,0 +1,325 @@
+
+        self.step("11a")
+        powerAdjustCapabilityStruct = await self.read_dem_attribute_expect_success(attribute="PowerAdjustmentCapability")
+        asserts.assert_greater_equal(len(powerAdjustCapabilityStruct.powerAdjustCapability), 1)
+        asserts.assert_equal(powerAdjustCapabilityStruct.cause,
+                             Clusters.DeviceEnergyManagement.Enums.PowerAdjustReasonEnum.kLocalOptimizationAdjustment)
+
+        self.step("12")
+        await self.send_power_adjustment_command(power=max_power,
+                                                 duration=min_duration,
+                                                 cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kGridOptimization)
+
+        # Wait 5 seconds for an event not to be reported
+        events_callback.wait_for_event_expect_no_report(5)
+
+        self.step("12a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPowerAdjustActive)
+
+        self.step("12b")
+        powerAdjustCapabilityStruct = await self.read_dem_attribute_expect_success(attribute="PowerAdjustmentCapability")
+        asserts.assert_greater_equal(len(powerAdjustCapabilityStruct.powerAdjustCapability), 1)
+        asserts.assert_equal(powerAdjustCapabilityStruct.cause,
+                             Clusters.DeviceEnergyManagement.Enums.PowerAdjustReasonEnum.kGridOptimizationAdjustment)
+
+        self.step("13")
+        await self.send_test_event_trigger_user_opt_out_local()
+
+        self.step("13a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPowerAdjustActive)
+
+        self.step("13b")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kLocalOptOut)
+
+        self.step("14")
+        await self.send_power_adjustment_command(power=max_power,
+                                                 duration=max_duration,
+                                                 cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization,
+                                                 expected_status=Status.ConstraintError)
+
+        self.step("15")
+        await self.send_test_event_trigger_user_opt_out_grid()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.PowerAdjustEnd)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kUserOptOut)
+
+        elapsed = datetime.datetime.now() - start
+        asserts.assert_less_equal(abs(elapsed.seconds - event_data.duration), 1)
+
+        # TODO: Do a better check on valid energyUse value.
+        # Value returned here is defined in DeviceEnergyManagementManufacturerImpl::GetEnergyUse()
+        asserts.assert_greater_equal(event_data.energyUse, 300)
```

Comment:
The test plan just says it has to be a valid value - so presumably >0?

Then we can remove the TODO?

Reply:
Done